icu4c/source/i18n/unitsdata.cpp

```diff
@@ -94,8 +95,247 @@ class ConversionRateDataSink : public ResourceSink {
     MaybeStackVector<ConversionRateInfo> *outVector;
 };
 
+UnitPreferenceMetadata::UnitPreferenceMetadata(const char *category, const char *usage,
+                                               const char *region, int32_t prefsOffset,
+                                               int32_t prefsCount, UErrorCode &status) {
+    this->category.append(category, status);
+    this->usage.append(usage, status);
+    this->region.append(region, status);
+    this->prefsOffset = prefsOffset;
+    this->prefsCount = prefsCount;
+}
+
+int32_t compareUnitPreferenceMetadata(const UnitPreferenceMetadata &a, const UnitPreferenceMetadata &b) {
+    int32_t cmp = uprv_strcmp(a.category.data(), b.category.data());
+    if (cmp == 0) { cmp = uprv_strcmp(a.usage.data(), b.usage.data()); }
+    if (cmp == 0) { cmp = uprv_strcmp(a.region.data(), b.region.data()); }
+    return cmp;
+}
+
+bool operator<(const UnitPreferenceMetadata &a, const UnitPreferenceMetadata &b) {
+    return compareUnitPreferenceMetadata(a, b) < 0;
+}
+
+int32_t compareUnitPreferenceMetadata(const UnitPreferenceMetadata &a, const UnitPreferenceMetadata &b,
+                                      bool *foundCategory, bool *foundUsage, bool *foundRegion) {
+    int32_t cmp = uprv_strcmp(a.category.data(), b.category.data());
+    if (cmp == 0) {
+        *foundCategory = true;
+        cmp = uprv_strcmp(a.usage.data(), b.usage.data());
+    }
+    if (cmp == 0) {
+        *foundUsage = true;
+        cmp = uprv_strcmp(a.region.data(), b.region.data());
+    }
+    if (cmp == 0) {
+        *foundRegion = true;
+    }
+    return cmp;
+}
+
+/**
+ * A ResourceSink that collects unit preferences information.
+ *
+ * This class is for use by ures_getAllItemsWithFallback.
+ */
+class UnitPreferencesSink : public ResourceSink {
+  public:
+    /**
+     * Constructor.
+     * @param outPrefs The vector to which UnitPreference instances are to be
+     * added. This vector must outlive the use of the ResourceSink.
+     * @param outMetadata  The vector to which UnitPreferenceMetadata instances
+     * are to be added. This vector must outlive the use of the ResourceSink.
+     */
+    explicit UnitPreferencesSink(MaybeStackVector<UnitPreference> *outPrefs,
+                                 MaybeStackVector<UnitPreferenceMetadata> *outMetadata)
+        : preferences(outPrefs), metadata(outMetadata) {}
+
+    /**
+     * Method for use by `ures_getAllItemsWithFallback`. Adds the unit
+     * preferences info that are found in `value` to the output vector.
+     *
+     * @param source This string must be "unitPreferenceData": the resource that
+     * this class supports reading.
+     * @param value The "unitPreferenceData" resource, containing unit
+     * preferences data.
+     * @param noFallback Ignored.
+     * @param status The standard ICU error code output parameter.
+     */
+    void put(const char *key, ResourceValue &value, UBool /*noFallback*/, UErrorCode &status) {
```

Comment:
This method contains a huge chunk of code. I prefer to split part of its code to a different function. Thus will make it simple to be read and to be understood from the code only.

Reply:
The nested for loops correspond to the JSON structure.  It makes sense when you put the two side-by-side.

Reply:
I've added more comments/documentation. I know I get a bit verbose with this... :-)

I considered pulling out a function, but I'm not sure this really improves things. A function name is more cryptic than a comment at the top of a block of code, and then either needs documentation anyway, or needs a very carefully considered name. (Or a name that feels too verbose for code.)

Let me know if the documentation is unclear, or comments could be improved? Or if the comments are simply too many. :-) (I go by the principle that reading the comment should let you skip blocks of code and not have to interpret them all - but the only people reading these comments might be those that need to dig in and understand the logic anyway?)

Reply:
I'm going to assume this is resolved.  @younies, if there is more you want from this, please file a follow-up PR.